### ISSUE TITLE
No trainable variables for real_nvp_default_template

### ISSUE THREAD

**Issue by modichirag** [tensorflow/probability#1439]
Hi, 

Configuration:
TF version '0.12.2'
TFP version '0.12.2'

I am trying to use real_nvp template in RealNVP bijector as given in the example in the docs
https://www.tensorflow.org/probability/api_docs/python/tfp/bijectors/RealNVP

sample() and log_prob() functions work but I do not see any trainable variables which are declared in the template.
```
tfd = tfp.distributions
tfb = tfp.bijectors

# A common choice for a normalizing flow is to use a Gaussian for the base
# distribution. (However, any continuous distribution would work.) E.g.,
nvp = tfd.TransformedDistribution(
    distribution=tfd.MultivariateNormalDiag(loc=[0., 0., 0.]),
    bijector=tfb.RealNVP(
        num_masked=2,
        shift_and_log_scale_fn=tfb.real_nvp_default_template(
            hidden_layers=[512, 512])))

x = nvp.sample()
nvp.log_prob(x)
nvp.log_prob([0.0, 0.0, 0.0])

print(nvp.trainable_variables)

>>>()
```

In fact the bijector itself does not seem to inherit any trainable variables

```
template = tfb.real_nvp_default_template(hidden_layers=[512, 512])
bijector = tfb.RealNVP(num_masked=1, shift_and_log_scale_fn=template)
bijector.forward([0.5, 1., 2.])
bijector.trainable_variables

>>> ()
```

Is this supposed to be so?

Thanks

**Comment by pointOfive** [comment]
I know this is not what you're asking about; but, `tfb.real_nvp_default_template` is not TF2.0 compatible, so, e.g., it won't work well with Keras.

I think the answer to your question is that you've not yet connected it with anything that causes the `.trainable_variables` to instantiate.  E.g.,

```
z = tfkl.Input(shape=(dim,), dtype=tf.float32)
p_z = tfd.TransformedDistribution(
       distribution = tfd.Sample(tfd.Normal(loc=3., scale=1.5), sample_shape=[dim]),
       bijector = tfb.MaskedAutoregressiveFlow(
       shift_and_log_scale_fn=tfb.AutoregressiveNetwork(
            params=2, hidden_units=[50]*1, #insufficient, also requires
            activation='relu', input_order='left-to-right')))]
print(p_z.trainable_variables)
p_z.log_prob(z)
print(p_z.trainable_variables)
```

**Comment by modichirag** [comment]
Hmm, interesting. I did not find it anywhere on the [docs](https://www.tensorflow.org/probability/api_docs/python/tfp/bijectors/real_nvp_default_template) that it is not TF2.0 compatible, unless I am missing something. Thanks for pointing this out!
(also I updated my original post to correct for TF version which was 2.4)

However for the current issue, I am not linking it with keras or keras.Models() anyways. I am just looking into extracting the gradients for trainable_variables with tf.Gradient.tape() and then pass them to the optimizer directly, bypassing Keras. 
Which is when I found this lack of variables. 

Regarding instantiating variables, I think I am doing that in the first code with
```
x = nvp.sample()
nvp.log_prob(x)
nvp.log_prob([0.0, 0.0, 0.0])
```
and in the second code with ```bijector.forward([0.5, 1., 2.])```
For instance, using my first code snippet and replacing ```shift_and_log_scale``` with ```tfb.AutoregressiveNetwork``` that you used instead of ```real_nvp_default_template``` works as expected, accessing trainable variables. 

Maybe it is cause is the compatibility issue that you mention?

**Comment by jecampagne** [comment]
Hello, is there any progress on RealNVP compatibility with TF v2, or a work around ?  I do not manage to get trainable variables but I am not an expert....
